### DISCUSSION TITLE
Graceful shutdown: how to communicate "shutting down" to active workers

### DISCUSSION THREAD

**Discussion by duncanb-q** [taskforcesh/bullmq#600] [Q&A]
To handle graceful shutdown, [this page](https://docs.bullmq.io/guide/workers/graceful-shutdown) indicates that `worker.close()` should be called.

That's great for not accepting new jobs, but our jobs are long-running (~hours).  We need a way to communicate 'the system is shutting down' so they can gracefully abort or pause themselves.  Something like:

```js
worker.forAllActiveJobs((j) => j.signal("shutdown"));
```

Is this possible in Bull? 

Thanks!

**Comment by manast** [comment]
This is actually a feature that I would like to implement short term, but I have not figured out a good API, do you have any suggestion?

**Reply by duncanb-q** [reply]
Extending your eventing model might not be a bad idea

```js
myWorker.on('shutdown', () => {
  // worker is shutting down
  // all job processing should attempt to shutdown cleanly
  // jobs should be re-queued an picked up by another eligible to another worker
  // the most common scenario for us here is "cluster restart", so we'd likely want 
  // a delay before having the jobs re-execute in another worker to avoid flip-flopping them 
  // during rolling restarts
});
```

FWIW, we also have a related required to support user-cancellation of a *specific* job: 

```js
myWorker.on('cancel', (jobId: string) => {
  // user requested cancellation of this job.  processing should attempt to shutdown cleanly
  // in this case job should NOT be re-queued
});
```

**Reply by manast** [reply]
I see. And when a job decides it has been cancelled, which status should it have, Completed or Failed? or maybe a new status "Cancelled" ?

**Reply by duncanb-q** [reply]
Ideally yes, cancelled since the job neither completed nor hit an error condition

**Comment by sixers** [comment]
Hey, any update on this?  `AbortController` / `AbortSignal` is quickly becoming a standard interface for cancelling promises, so it probably makes sense to implement it with them. It's already used in various async APIs in node (e.g. `fs`, fetch (`node-fetch` or new bundled `undici` implementation in nodejs since 17.5)). 

`Job` instance could have a signal property, which could be checked by job code & passed to other async calls.

Example:

```javascript
async function handleJob(job) {
  const { signal, data } = job;

  // check signal at the beginning
  if (signal.aborted === true) {
    console.log("Job cancelled");
    return;
  }

  // pass signal to other calls
  const response = await fetch(data.url, { signal });
  const output = await response.json();
  // ... do something else
  // ... do some other async work

  // check signal again
  if (signal.aborted) {
    console.log("Job cancelled");
    return;
  }

  return output;
}
```

**Comment by manast** [comment]
We have a cancel mechanism in the pro version, however, is based on observables. In the future, if AbortController gets more traction we could implement it too. 

Btw, the important feature in my opinion of the abort controller would be the "abort" listener, which did not get depicted in your code example, in general, it should not be needed to synchronously test for ```signal.aborted``` in the job itself.

**Comment by sixers** [comment]
@manast How do you think `abort` listener should be used in this context?
Would you be open to a pull request which implements cancellation via `AbortController`?

**Reply by manast** [reply]
You use the abort event to cancel your promise:
```ts
return new Promise((resolve, reject) => {
  signal.on('abort', reject);
});
```

**Reply by sixers** [reply]
Yeah, I guess it will depend on the exact implementation of the job handler. In some cases it might make sense to use the signal event handler to cancel. Not a silver bullet though, because it doesn't guarantee that whatever is started in the `Promise` is actually cancelled. In general, I think it should be up to job handler to decide how to respond to the cancellation.

Synchronous checks might still be required, because not every downstream call will support `AbortSignal` and propagate cancellation (e.g. Prisma doesn't support cancellation at the moment).

Consider the following code sample:

```javascript

async function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function run1(n, {ms = 0, signal}) {
  for (let i = 0; i < n; i++) {
    console.log("run1", i);
    await sleep(ms);
  }
}

async function test1() {
  console.log("Running test1");
  const abort1 = new AbortController();
  const promise1 = run1(10, {ms: 1000, signal: abort1.signal});
  await sleep(5000);
  console.log("cancel");
  abort1.abort();
  console.log("cancelled");
  await promise1;
  console.log("returned");
}

function run2(n, {ms = 0, signal}) {
  return new Promise((resolve, reject) => {
    signal?.addEventListener("abort", reject);
    (async () => {
       for (let i = 0; i < n; i++) {
        console.log("run2", i);
        await sleep(ms);
      }
    })().then(resolve);
  })
} 

async function test2() {
  console.log("Running test2");
  const abort1 = new AbortController();
  const promise1 = run2(10, {ms: 1000, signal: abort1.signal});
  await sleep(5000);
  console.log("cancel");
  abort1.abort();
  console.log("cancelled");
  try {
    
  
  await promise1;
  console.log("returned");
    } catch (e) {
    if (e.type === 'abort') {
    console.log("aborted")
      } else {
      console.error(e);
      }
    }
}



function run3(n, {ms = 0, signal}) {
  return new Promise((resolve, reject) => {
    (async () => {
       for (let i = 0; i < n; i++) {
         if (signal.aborted) {
           const error = new Error();
           error.type = "abort";
           return reject(error);
         }
        console.log("run3", i);
        await sleep(ms);
      }
    })().then(resolve);
  })
} 

async function test3() {
  console.log("Running test3");
  const abort1 = new AbortController();
  const promise1 = run3(10, {ms: 1000, signal: abort1.signal});
  await sleep(5000);
  console.log("cancel");
  abort1.abort();
  console.log("cancelled");
  try {
    
  await promise1;
  console.log("returned");
    } catch (e) {
    if (e.type === 'abort') {
    console.log("aborted")
      } else {
      console.error(e);
      }
    }
}

async function testall() {
  await Promise.allSettled([test1(), sleep(11 * 1000)]);
  await Promise.allSettled([test2(), sleep(11 * 1000)]);
  await Promise.allSettled([test3(), sleep(11 * 1000)]);
}

testall();
```

Output:

```
 testall();
Running test1
run1 0
Promise {
  <pending>,
  [Symbol(async_id_symbol)]: 48,
  [Symbol(trigger_async_id_symbol)]: 2,
  [Symbol(destroyed)]: { destroyed: false }
}
 run1 1
run1 2
run1 3
run1 4
cancel
cancelled
run1 5
run1 6
run1 7
run1 8
run1 9
returned
Running test2
run2 0
run2 1
run2 2
run2 3
run2 4
cancel
cancelled
aborted
run2 5
run2 6
run2 7
run2 8
run2 9
Running test3
run3 0
run3 1
run3 2
run3 3
run3 4
cancel
cancelled
aborted

```


I think the most important thing is to actually expose the `AbortSignal` (and perhaps `AbortController` as well) as property on `Job` or as an argument to `Processor` , and implement a `cancelJob` method in Queue which should trigger the `AbortController.abort()`.

I could try to implement this, would be be willing to review a PR which implements this feature?

**Reply by manast** [reply]
The only feature that can easily be implemented is the cancelation due to TTL. A generic cancel method requires a communication mechanism with the worker that is actually processing a job. As I mentioned earlier, I am open to a feature like this but I would rather wait until it becomes more established. As you say there are many libraries that do not support it yet, furthermore the AbortController is only available in version 15 and above.
